google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/ManagedCloudSdkUpdateService.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2018 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.intellij.appengine.sdk;
+
+import com.google.cloud.tools.intellij.GctFeature;
+import com.google.cloud.tools.intellij.service.PluginInfoService;
+import com.google.common.annotations.VisibleForTesting;
+import com.intellij.notification.Notification;
+import com.intellij.openapi.components.ServiceManager;
+import java.awt.event.ActionListener;
+import java.time.Clock;
+import java.util.Timer;
+import java.util.TimerTask;
+
+/** Schedules and manages automatic updates for {@link ManagedCloudSdkService}. */
+public class ManagedCloudSdkUpdateService {
+  // one week.
+  @VisibleForTesting static final long SDK_UPDATE_INTERVAL_MS = 1000 * 60 * 60 * 24 * 7;
+  // 20 seconds to show notification and then proceed with update.
+  private static final int UPDATE_NOTIFICATION_DELAY_MS = 1000 * 20;
+
+  @VisibleForTesting
+  static final String SDK_UPDATER_THREAD_NAME = "managed-google-cloud-sdk-updates";
+
+  static ManagedCloudSdkUpdateService getInstance() {
+    return ServiceManager.getService(ManagedCloudSdkUpdateService.class);
+  }
+
+  private Timer updateTimer;
+  private TimerTask sdkUpdateTask;
+
+  void activate() {
+    if (!CloudSdkServiceUserSettings.getInstance().isAutomaticUpdateEnabled()) {
+      return;
+    }
+
+    if (updateTimer == null) {
+      updateTimer = new Timer(SDK_UPDATER_THREAD_NAME);
+    }
+    // cancel tasks from previous activation.
+    if (sdkUpdateTask != null) {
+      sdkUpdateTask.cancel();
+    }
+
+    sdkUpdateTask =
+        new TimerTask() {
+          @Override
+          public void run() {
+            doUpdate();
+          }
+        };
+    long lastTimeOfUpdate =
```

Comment:
perhaps extract all of the below (up to line 75) into a well named method that indicates what all the calculations do for readability? Also with a javadoc